language_service/src/hover.rs

```diff
@@ -25,67 +29,447 @@ pub(crate) fn get_hover(
+                            Some(contents_from_name(&ident.name, &self.get_type_name(pat.id)));
+                        self.start = ident.span.lo;
+                        self.end = ident.span.hi;
+                    } else if let Some(ty) = anno {
+                        self.visit_ty(ty);
+                    }
+                }
+                _ => walk_pat(self, pat),
+            }
+        }
+    }
+
+    fn visit_expr(&mut self, expr: &'_ ast::Expr) {
+        if span_contains(expr.span, self.offset) {
+            match &*expr.kind {
+                ast::ExprKind::Field(_, field) if span_contains(field.span, self.offset) => {
+                    self.contents = Some(contents_from_name(
+                        &field.name,
+                        &self.get_type_name(expr.id),
+                    ));
+                    self.start = field.span.lo;
+                    self.end = field.span.hi;
+                }
+                _ => walk_expr(self, expr),
+            }
+        }
+    }
+
+    fn visit_path(&mut self, path: &'_ ast::Path) {
+        if span_contains(path.span, self.offset) {
+            let res = self.compilation.unit.ast.names.get(path.id);
+            if let Some(res) = res {
+                match &res {
+                    resolve::Res::Item(item_id) => {
+                        if let Some(item) = find_item(self.compilation, item_id) {
+                            self.contents = match &item.kind {
+                                hir::ItemKind::Callable(decl) => {
+                                    Some(self.contents_from_hir_call_decl(decl, &item.doc))
+                                }
+                                hir::ItemKind::Namespace(_, _) => {
+                                    panic!(
+                                        "Reference node should not refer to a namespace: {}",
+                                        path.id
+                                    )
+                                }
+                                hir::ItemKind::Ty(ident, udt) => {
+                                    Some(contents_from_hir_udt(ident, udt))
+                                }
+                            };
+                            self.start = path.span.lo;
+                            self.end = path.span.hi;
+                        }
+                    }
+                    resolve::Res::Local(node_id) => {
+                        self.contents = Some(contents_from_name(
+                            &print_path(path),
+                            &self.get_type_name(*node_id),
+                        ));
+                        self.start = path.span.lo;
+                        self.end = path.span.hi;
+                    }
+                    _ => {}
+                };
+            }
+        }
+    }
+}
+
+impl HoverVisitor<'_> {
+    fn contents_from_ast_call_decl(&mut self, decl: &ast::CallableDecl, doc: &Rc<str>) -> String {
```

Comment:
These seem like general useful formatting functions that we'll probably want to extract out to a common module. Other features (for example [completion item details](https://code.visualstudio.com/docs/editor/intellisense)) will probably need them.

So maybe add a `display.rs` module and move all your `format_xyz()` functions there? I do see that some of them will need to take a `find_item` callback which you can pass in from here.

Reply:
I'm struggling to find a way to make this work with the use of the `terms` and `names` maps needed to resolve types. I'm also convinced that this ought to be done in an OOP way rather than through passing around callback functions to a lot of functions when it would only be used in two places, but that doesn't mean they have to live in the `HoverVisitor` type. Do you have any further suggestions here?

Reply:
Thanks for your help with this!